third_party/move/move-vm/runtime/src/runtime_type_checks.rs

```diff
@@ -0,0 +1,683 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{frame_type_cache::FrameTypeCache, interpreter::Stack, loader::Resolver};
+use move_binary_format::{
+    errors::*,
+    file_format::{Ability, AbilitySet, Bytecode},
+};
+use move_core_types::vm_status::StatusCode;
+use move_vm_types::{loaded_data::runtime_types::Type, values::Locals};
+
+pub(crate) trait RuntimeTypeCheck {
+    fn pre_execution_type_stack_transition(
+        local_tys: &[Type],
+        locals: &Locals,
+        _ty_args: &[Type],
+        _resolver: &Resolver,
+        operand_stack: &mut Stack,
+        instruction: &Bytecode,
+    ) -> PartialVMResult<()>;
+
+    fn post_execution_type_stack_transition(
+        local_tys: &[Type],
+        ty_args: &[Type],
+        resolver: &Resolver,
+        operand_stack: &mut Stack,
+        ty_cache: &mut FrameTypeCache,
+        instruction: &Bytecode,
+    ) -> PartialVMResult<()>;
+}
+
+/// Paranoid type checks to perform before instruction execution.
+///
+
+fn verify_pack<'a>(
+    operand_stack: &mut Stack,
+    field_count: u16,
+    field_tys: impl Iterator<Item = &'a Type>,
+    output_ty: Type,
+) -> PartialVMResult<()> {
+    let ability = output_ty.abilities()?;
+
+    // If the struct has a key ability, we expect all of its field to have store ability but not key ability.
+    let field_expected_abilities = if ability.has_key() {
+        ability
+            .remove(Ability::Key)
+            .union(AbilitySet::singleton(Ability::Store))
+    } else {
+        ability
+    };
+    for (ty, expected_ty) in operand_stack
+        .popn_tys(field_count)?
+        .into_iter()
+        .zip(field_tys)
+    {
+        // Fields ability should be a subset of the struct ability because abilities can be weakened but not the other direction.
+        // For example, it is ok to have a struct that doesn't have a copy capability where its field is a struct that has copy capability but not vice versa.
+        ty.paranoid_check_abilities(field_expected_abilities)?;
+        ty.paranoid_check_eq(expected_ty)?;
+    }
+
+    operand_stack.push_ty(output_ty)
+}
+
+pub(crate) struct NullRuntimeTypeCheck;
+pub(crate) struct FullRuntimeTypeCheck;
```

Comment:
nit: move `FullRuntimeTypeCheck` to trait impl? Also add brief documentation? Space between struct definitions?

Reply:
Disagree with this - prefer to have the two choices next to each other (before the implementation). As for documentation - I think the names are self documenting :)

Reply:
Don't have strong opinion on this, so fine with me